lib/modules/manager/gradle/parser.ts

```diff
@@ -1,1208 +1,458 @@
+  .alt(
+    qStringValue,
+    q.tree({
+      type: 'wrapped-tree',
+      maxDepth: 1,
+      startsWith: '(',
+      endsWith: ')',
+      search: q.begin<Ctx>().join(qStringValue).end(),
+    })
+  )
+  .handler((ctx) => storeInTokenMap(ctx, 'pluginName'))
+  .sym('version')
+  .alt(qTemplateString, qVariableAccessIdentifier)
+  .handler((ctx) => storeInTokenMap(ctx, 'version'))
+  .handler(handlePlugin)
+  .handler(cleanupTempVars);
+
+// mavenCentral()
+// mavenCentral { ... }
+const qPredefinedRegistries = q
+  .sym(regEx(`^(?:${Object.keys(REGISTRY_URLS).join('|')})$`), storeVarToken)
+  .alt(
+    q.tree({
+      type: 'wrapped-tree',
+      startsWith: '(',
+      endsWith: ')',
+      search: q.begin<Ctx>().end(),
+    }),
+    q.tree({
+      type: 'wrapped-tree',
+      startsWith: '{',
+      endsWith: '}',
+    })
+  )
+  .handler((ctx) => storeInTokenMap(ctx, 'registryUrl'))
+  .handler(handlePredefinedRegistryUrl)
+  .handler(cleanupTempVars);
+
+// maven(url = uri("https://foo.bar/baz"))
+// maven { name = some; url = "https://foo.bar/${name}" }
+const qCustomRegistryUrl = q
+  .sym<Ctx>('maven')
+  .alt(
+    q.tree<Ctx>({
+      type: 'wrapped-tree',
+      maxDepth: 1,
+      startsWith: '(',
+      endsWith: ')',
+      search: q
+        .begin<Ctx>()
+        .opt(q.sym<Ctx>('url').op('='))
+        .alt(
+          q.sym<Ctx>('uri').tree({
+            maxDepth: 1,
+            search: q.alt(qTemplateString, qVariableAccessIdentifier),
+          }),
+          q.alt(qTemplateString, qVariableAccessIdentifier)
+        )
+        .end(),
+    }),
+    q.tree({
+      type: 'wrapped-tree',
+      maxDepth: 1,
+      startsWith: '{',
+      endsWith: '}',
+      search: q.alt(
+        q
+          .sym<Ctx>('name')
+          .opt(q.op('='))
+          .alt(qTemplateString, qVariableAccessIdentifier)
+          .handler((ctx) => storeInTokenMap(ctx, 'name')),
+        q
+          .sym<Ctx>('url')
+          .opt(q.op('='))
+          .alt(
+            q.sym<Ctx>('uri').tree({
+              maxDepth: 1,
+              search: q.alt(qTemplateString, qVariableAccessIdentifier),
+            }),
+            q.alt(qTemplateString, qVariableAccessIdentifier)
+          ),
+        q.sym<Ctx>('setUrl').tree({
+          maxDepth: 1,
+          startsWith: '(',
+          endsWith: ')',
+          search: q
+            .begin<Ctx>()
+            .alt(qTemplateString, qVariableAccessIdentifier)
+            .end(),
+        })
+      ),
+    })
+  )
+  .handler((ctx) => storeInTokenMap(ctx, 'registryUrl'))
+  .handler(handleCustomRegistryUrl)
+  .handler(cleanupTempVars);
+
+const qVersionCatalogVersion = q
+  .op<Ctx>('.')
+  .alt(
+    // library("kotlin-reflect", "org.jetbrains.kotlin", "kotlin-reflect").versionRef("kotlin")
+    q.sym<Ctx>('versionRef').tree({
+      maxDepth: 1,
+      startsWith: '(',
+      endsWith: ')',
+      search: q.begin<Ctx>().join(qStringValueAsSymbol).end(),
```

Comment:
Why not to use string matcher? I know it breaks the test, though not sure why

```suggestion
      search: q.begin<Ctx>().str(storeVarToken).end(),
```

Reply:
I'd be worried, if the test wouldn't fail here :wink:

The expression this relates to is, e.g. `versionRef("kotlin")`. `str(storeVarToken)` would store "kotlin" as a string value, although the value itself references a variable. By storing it as a symbol here, the handler function will immediately get the token with the correct type associated.

The alternative would be to remember if `version` or `versionRef` was parsed and then, in the handler, do the conversion. This is how it is done in the current parser: 
https://github.com/renovatebot/renovate/blob/6f051f3ae7e8f63dfe7481dc5f82d30a5b09ab55/lib/modules/manager/gradle/parser.ts#L392-L401

Reply:
Thanks, I'm okay with your implementation